### PR TITLE
fix(controller): Workflow hangs indefinitely during ContainerCreating if the Pod or Node unexpectedly dies

### PR DESCRIPTION
**Pull request by sarabala1979**
Signed-off-by: Saravanan Balasubramanian <sarabala1979@gmail.com>

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -975,7 +975,9 @@ func (woc *wfOperationCtx) podReconciliation(ctx context.Context) error {
 
 			// If the node is pending and the pod does not exist, it could be the case that we want to try to submit it
 			// again instead of marking it as an error. Check if that's the case.
-			if node.Pending() {
+			// Added `ContainerCreating` condition to avoid workflow stuck in running state with the pod deleted in ContainerCreating
+			// https://github.com/argoproj/argo-workflows/issues/5543
+			if node.Pending() && !strings.Contains(node.Message, "ContainerCreating") {
```
Comment (alexec) [inline workflow/controller/operator.go:980]: Why is the node not pending if the pod is creating? or is the node running? I wonder if something is wrong with `inferPodPhase`, take a look at this:

https://github.com/argoproj-labs/argo-dataflow/blob/main/controllers/infer.go#L40

Comment (alexec) [reply workflow/controller/operator.go:980]: I think a pod can have `phase=Running`, but has containers are in `ContainerCreating` status. 

I think a deep-dive into pod status?

**Comment by sarabala1979** [comment]
`Pod` will be in the `pending` status if any of the  container is not in `containerReady` state. Code is setting pending status if container in `waiting` state.

https://github.com/argoproj/argo-workflows/blob/4e450e250168e6b4d51a126b784e90b11a0162bc/workflow/controller/operator.go#L1118-L1133
Cont>     
>     Name:           my-nginx-9d5677d94-g44l6 Namespace:      default Node: kubenode1/10.1.88.22 Start Time:     Tue, 06 Mar 2018 08:24:13
> +0000 Labels:         
pod-template-hash=581233850
>                     run=my-nginx Annotations:    <none> 
**Status:         Pending** 
IP: Controlled By:  ReplicaSet/my-nginx-9d5677d94 Containers: 
> my-nginx:
>         Container ID:
>         Image:          nginx
>         Image ID:
>         Port:           80/TCP
**>         State:          Waiting
>           Reason:       ContainerCreating**
>         Ready:          False
>         Restart Count:  0
>         Environment:    <none>roller is checking only pod status in code.
```

**Comment by sarabala1979** [comment]
@alexec  Do you think we can move nodestatus to  `nodeRunning` when the container state is  is in `waiting`? Currently code is keeping `nodepending`.  
https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle

```

Container states
As well as the phase of the Pod overall, Kubernetes tracks the state of each container inside a Pod. You can use container lifecycle hooks to trigger events to run at certain points in a container's lifecycle.

Once the scheduler assigns a Pod to a Node, the kubelet starts creating containers for that Pod using a container runtime. There are three possible container states: Waiting, Running, and Terminated.

To check the state of a Pod's containers, you can use kubectl describe pod <name-of-pod>. The output shows the state for each container within that Pod.

Each state has a specific meaning:

Waiting
If a container is not in either the Running or Terminated state, it is Waiting. A container in the Waiting state is still running the operations it requires in order to complete start up: for example, pulling the container image from a container image registry, or applying Secret data. When you use kubectl to query a Pod with a container that is Waiting, you also see a Reason field to summarize why the container is in that state.

```

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -975,7 +975,9 @@ func (woc *wfOperationCtx) podReconciliation(ctx context.Context) error {
 
 			// If the node is pending and the pod does not exist, it could be the case that we want to try to submit it
 			// again instead of marking it as an error. Check if that's the case.
-			if node.Pending() {
+			// Added node.Message condition to avoid workflow stuck in running state with the pod deleted in ContainerCreating
+			// https://github.com/argoproj/argo-workflows/issues/5543
+			if node.Pending() && node.Message != "" {
```
Comment (sarabala1979) [inline workflow/controller/operator.go:980]: node.Message will be updated if Pod scheduled and container is starting

File: workflow/controller/operator.go [hunk 3]
```diff
@@ -975,7 +975,8 @@ func (woc *wfOperationCtx) podReconciliation(ctx context.Context) error {
 
 			// If the node is pending and the pod does not exist, it could be the case that we want to try to submit it
 			// again instead of marking it as an error. Check if that's the case.
-			if node.Pending() {
+			// Node will be in pending state without Pod create if Node is waiting for Synchronize lock
+			if node.Pending() && !node.IsPodCreated() {
```
Comment (sarabala1979) [inline workflow/controller/operator.go:979]: There is one valid scenario where node status will be `pending` without pod creation. It is the template that is waiting for sync lock to acquire.

**Comment by alexec** [comment]
`pod.status.phase` is not a good way to understand the state of the pod.

So what is better?

@sarabala1979 I'm not sure what to do right now, I did wonder if gitops-engine is correct:

https://github.com/argoproj/gitops-engine/blob/master/pkg/health/health_pod.go#L14

But no, not correct.

What about `kubect get pod` - how does that determine `STATUS`?

```
 k get pod
NAME                     READY   STATUS    RESTARTS   AGE
minio-657cbb9d6d-2225w   1/1     Running   0          30h
nats-0                   1/1     Running   0          19h
stan-0                   1/1     Running   0          19h
```

I know that we can be running, but status is things like

* CrashloopBackoff
* NotReady
* ErrImagePull

Do you know where to find the source code?

**Comment by alexec** [comment]
To make this worse - a "waiting" container can be in error state and not recoverable!

File: workflow/controller/operator.go [hunk 4]
```diff
@@ -976,9 +976,10 @@ func (woc *wfOperationCtx) podReconciliation(ctx context.Context) error {
 			// If the node is pending and the pod does not exist, it could be the case that we want to try to submit it
 			// again instead of marking it as an error. Check if that's the case.
 			// Node will be in pending state without Pod create if Node is waiting for Synchronize lock
-			if node.Pending() && !node.IsPodCreated() {
+			if node.Pending() && node.GetPendingReason() == wfv1.WaitingForSyncLock {
```
Comment (alexec) [inline workflow/controller/operator.go:979]: should this be `||` not `&&`?

Comment (sarabala1979) [reply workflow/controller/operator.go:979]: should be &&

Comment (sarabala1979) [reply workflow/controller/operator.go:979]: The node should not fail/error if the node is in a pending state with WaitforSyncLock.

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 5]
```diff
@@ -1279,6 +1285,14 @@ func (n Nodes) Find(f func(NodeStatus) bool) *NodeStatus {
 	return nil
 }
 
+func (ns NodeStatus) GetPendingReason() NodePendingReason {
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1288]: maybe use `GetReason`?

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 6]
```diff
@@ -93,6 +93,12 @@ const (
 	VolumeClaimGCOnSuccess    VolumeClaimGCStrategy = "OnWorkflowSuccess"
 )
 
+type NodePendingReason string
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:96]: maybe `NodePhaseReason`?

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 7]
```diff
@@ -1279,6 +1285,14 @@ func (n Nodes) Find(f func(NodeStatus) bool) *NodeStatus {
 	return nil
 }
 
+func (ns NodeStatus) GetPendingReason() NodePendingReason {
+	// If node is waiting for synchronize lock, Pod will not be created in this scenario
+	if ns.SynchronizationStatus != nil && ns.SynchronizationStatus.Waiting != "" {
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1290]: unit test?

File: workflow/controller/operator.go (hunk 4 above)
Comment (alexec) [reply workflow/controller/operator.go:979]: are you sure? `&&` means that any node that is pending and not waiting for lock may be marked as error

Comment (sarabala1979) [reply workflow/controller/operator.go:979]: yes. I am sure.  There is another scenario (pod. is not updated to informer) which will cover in `recentlyStarted`

Comment (alexec) [reply workflow/controller/operator.go]: @simster7 - can I request another opinion on this? I feel like I'm going mad.

**Review by alexec — APPROVED** [review]

**Pull request merged**